retis/src/core/filters/filters.rs

```diff
@@ -60,81 +60,161 @@ pub(crate) fn register_filter_handler(
+            Filter::Packet(_, bf) => bf.0,
+            _ => {
+                warn!(
+                    "Found invalid type while retrieving the filter, defaulting to match all ..."
+                );
+                get_default_filter()
+            }
+        }
+    } else {
+        get_default_filter()
+    };
+
+    let filter: &[libbpf_sys::bpf_insn] = unsafe {
+        std::slice::from_raw_parts(
+            f.as_slice().as_ptr() as *const libbpf_sys::bpf_insn,
+            f.len() / std::mem::size_of::<libbpf_sys::bpf_insn>(),
+        )
+    };
+
+    filter.to_vec()
+}
+
+fn copy_ptr_to_vec<T>(ptr: *mut T, count: usize) -> Vec<T> {
+    let mut vec = Vec::with_capacity(count);
+    unsafe {
+        std::ptr::copy(ptr, vec.as_mut_ptr(), count);
+        vec.set_len(count);
+    }
+    vec
+}
+
+// Rewrites the instructions, indexes all the pseduo calls, fixes
+// them up along with the relevant .BTF.ext.
 pub(crate) unsafe extern "C" fn fixup_filter_load_fn(
     prog: *mut libbpf_sys::bpf_program,
-    _opts: *mut libbpf_sys::bpf_prog_load_opts,
+    opts: *mut libbpf_sys::bpf_prog_load_opts,
     _cookie: ::std::os::raw::c_long,
 ) -> std::os::raw::c_int {
-    for magic in [packet_filter_uapi::FILTER_L2, packet_filter_uapi::FILTER_L3] {
-        let filter = get_filter(magic);
-
-        let f = if let Some(f) = filter {
-            match f {
-                Filter::Packet(_, bf) => bf.0,
-                // fail if non packet filter is encountered
-                _ => return -1,
-            }
-        } else {
-            let mut default_filter = eBpfProg::new();
-            default_filter.add(eBpfInsn::mov32(MovInfo::Imm {
-                dst: BpfReg::R0,
-                imm: 0x40000_i32,
-            }));
-            default_filter.to_bytes()
-        };
+    let filter_types = [packet_filter_uapi::L2, packet_filter_uapi::L3];
+    let mut pseudo_calls: Vec<usize> = Vec::new();
+    let mut placeholder_calls: Vec<(usize, _)> = Vec::new();
+
+    let insns = unsafe {
+        std::slice::from_raw_parts(
+            libbpf_sys::bpf_program__insns(prog),
+            libbpf_sys::bpf_program__insn_cnt(prog) as usize,
+        )
+    };
 
-        let filter: &[libbpf_sys::bpf_insn] = unsafe {
-            std::slice::from_raw_parts(
-                f.as_slice().as_ptr() as *const libbpf_sys::bpf_insn,
-                f.len() / std::mem::size_of::<libbpf_sys::bpf_insn>(),
-            )
-        };
+    for (pos, insn) in insns.iter().enumerate() {
+        let imm = insn.imm as u32;
 
-        let (insns, insns_cnt) = unsafe {
-            (
-                libbpf_sys::bpf_program__insns(prog),
-                libbpf_sys::bpf_program__insn_cnt(prog),
-            )
-        };
+        // 0xBAD2310 is used to mark instructions that failed
+        // relocation:
+        // 1: (85) call unknown#195896080
+        // invalid func unknown#195896080
+        // this is skipped as this is intended and handled.
```

Comment:
Can you add this is done specifically by `libbpf` and not us?

Reply:
"failed relocation" implies it (also the log is from libbpf), but I'm ok with being more explicit.

Reply:
uhm, tbf the log was not the intended one.
While at it, let's replace that and solve both.